From a04af2a5b20951532dab00a0c13c65d95f15c738 Mon Sep 17 00:00:00 2001
From: Aniket Randive <arandive@qti.qualcomm.com>
Date: Mon, 8 Sep 2025 14:55:43 +0530
Subject: [PATCH] i2c: i2c-qcom-geni: Add data transfer support on multi GPII

Enhance the I2C driver to support data transfer on multiple GPIIs
in GSI mode for CCU QUP. Previously, the driver only supported a
single GPII. This update allows a maximum of 3 GPIIs for a single
serial engine, improving flexibility and performance.

Additionally, for CCU QUP, the client will update the following
required configurations in the first 8 bytes of message buffer
of struct i2c_msg during data transfer:

| Buffer | Field         | Bits   | Description                       |
|---------------------------------------------------------------------|
| Byte 0 | GPII Number   | 0-4 | Represents the GPII no. According    |
|        |               |        | to this GPII no, I2C driver will  |
|        |               |        | send the received message/packet  |
|        |               |        | on sync ch0 or sync ch1 or regular|
|        |               |        | channel.                          |
|        |------------------------------------------------------------|
|        | Async transfer|        | In Async mode, I2C driver will    |
|        | mode          | 5      | return dummy value after queuing a|
|        |               |        | data transfer. The QUP driver     |
|        |               |        | notifies the client via poll API  |
|        |               |        | after a GSI completion interrupt  |
|        |               |        | for a write transfer. Async mode  |
|        |               |        | is only applicable for write xfer.|
|        |------------------------------------------------------------|
|        | Priority mode | 6      | Used only when client wants to    |
|        |               |        | send high priority data transfer  |
|        |               |        | on regular channel.               |
|        |------------------------------------------------------------|
|        | NOP Frame     | 7      | Used only when SW wants to NOP the|
|        |               |        | FRAME(s).                         |
|---------------------------------------------------------------------|
| Byte1	 | Insert delay  | 0      | Insert predefined delay before I2C|
|        | before        |        | transaction.                      |
|        |------------------------------------------------------------|
|        | Insert delay  | 1      | Insert predefined delay after I2C |
|        | after         |        | transaction.                      |
|        |------------------------------------------------------------|
|        | Ch last       | 2      | Indicates last packet(TX/RX) of a |
|        |               |        | current FRAME.                    |
|        |------------------------------------------------------------|
|        | Ccu last      | 3      | This bit indicates end of transfer|
|        |               |        | for single SYNC request.          |
|        |------------------------------------------------------------|
|        | ARB lost Retry| 4-5    | Re-transmit count for Arb lost in |
|        |               |        | Address phase.                    |
|        |------------------------------------------------------------|
|        |    -          | 6      | Reserved                          |
|        |------------------------------------------------------------|
|        |    -          | 7      | Reserved                          |
|---------------------------------------------------------------------|
| Byte2  | Message len   | 0-5    | TX sub-packet length for multiple |
|        |               |        | Random Write feature. Nonzero     |
|        |               |        | value will enable Multi-Random    |
|        |               |        | write feature.                    |
|        |------------------------------------------------------------|
|        |    -          | 6-7    | Reserved                          |
|---------------------------------------------------------------------|
| Byte3	 |    -          | 0-7    | Reserved                          |
|---------------------------------------------------------------------|
| Byte4  | Frame cookie  | 0-7    | Frame cookie (lower 0-7 bits)     |
|---------------------------------------------------------------------|
| Byte5  | Frame cookie  | 0-7    | Frame cookie (lower 15-8 bits)    |
|---------------------------------------------------------------------|
| Byte6  | Frame cookie	 | 0-7    | Frame cookie (higher 23-16 bits)  |
|---------------------------------------------------------------------|
| Byte7  | Frame cookie  | 0-7    | Frame cookie (higher 31-24 bits)  |
|---------------------------------------------------------------------|

Hence, data buffer parse support has been added.

Change-Id: Ia03b6a42c3e8ef04d8be2f5beea4258916078f85
Signed-off-by: Aniket Randive <arandive@qti.qualcomm.com>
---
 drivers/i2c/busses/i2c-qcom-geni.c | 334 ++++++++++++++++++++++++-----
 1 file changed, 283 insertions(+), 51 deletions(-)

diff --git a/drivers/i2c/busses/i2c-qcom-geni.c b/drivers/i2c/busses/i2c-qcom-geni.c
index 805e8df74274..ea8e0002f29e 100644
--- a/drivers/i2c/busses/i2c-qcom-geni.c
+++ b/drivers/i2c/busses/i2c-qcom-geni.c
@@ -97,9 +97,86 @@ struct geni_i2c_desc {
 	bool no_dma_support;
 	unsigned int tx_fifo_depth;
 	int (*geni_se_switch_state)(struct device *dev, bool state);
+	bool is_ccu_qup;
 };
 
-static int geni_i2c_runtime_resume(struct device *dev);
+/* CCU FRAME */
+#define CCU_DATA_OFFSET		8
+#define CCU_GPII_MASK		GENMASK(3, 0)
+#define CCU_ARB_LOST_MASK	GENMASK(16, 13)
+#define CCU_ARB_LOST_SHFT	13
+#define CCU_PKT_LEN_MASK	GENMASK(21, 16)
+#define CCU_PKT_LEN_SHFT	16
+#define CCU_FRAME_COOKIE_MASK	GENMASK(63, 32)
+#define CCU_FRAME_COOKIE_SHFT	32
+
+#define CCU_MAX_GPII		3
+#define REGULAR_GPII		2
+#define MAX_NUM_MSGS		4096
+#define CCU_GET_GPII(val)	(val & CCU_GPII_MASK)
+#define CCU_GET_SYNC(val)	(val & BIT(5))
+#define CCU_GET_PRIO(val)	(val & BIT(6))
+#define CCU_GET_NOP_FRAME(val)	(val & BIT(7))
+#define CCU_GET_INSERT_DELAY_BEFORE(val)	(val & BIT(8))
+#define CCU_GET_INSERT_DELAY_AFTER(val)		(val & BIT(9))
+#define CCU_GET_CH_LAST(val)		(val & BIT(11))
+#define CCU_GET_LAST(val)		(val & BIT(12))
+#define CCU_GET_ARB_LOST_CNT(val)	((val & CCU_ARB_LOST_MASK) >> CCU_ARB_LOST_SHFT)
+#define CCU_GET_PKT_LEN(val)		((val & CCU_PKT_LEN_MASK) >> CCU_PKT_LEN_SHFT)
+#define CCU_GET_FRAME_COOKIE(val)	((val & CCU_FRAME_COOKIE_MASK) >> CCU_FRAME_COOKIE_SHFT)
+
+/* struct ccu_gpii_info - represents the GPII information
+ * @type: indicate current gpii is regular or sync ch0 or sync ch1
+ * @gpii_mode_change: flag to indicate the mode change
+ */
+struct ccu_gpii_info {
+	int type; /* Regular or Sync channel */
+	bool gpii_mode_change;
+};
+
+/*
+ * struct ccu_msg_config - used to store the configuration of single frame.
+ * @gpii_no: represents the GPII no on which transfer need to send
+ * @sync: represents the synchronous or asynchornous transfer
+ * @priority: represents high priority data transfer
+ * @nop_last: when SW wants to NOP the FRAME(s)
+ * @pre_delay: insert predefined delay before I2C transaction
+ * @post_delay: insert predefined delay after I2C transaction
+ * @pkt_len: tx sub-packet length for multiple random write
+ * @arb_lost_retry: re-transmit count for arb lost in address phase
+ * @ch_last: indicates last packet (TX & RX) of a current FRAME
+ * @ccu_last: indicates end of transfer for single SYNC request
+ * @frame_cookie: indicate unique frame cookie for transfer.
+ */
+struct ccu_msg_config {
+	int gpii_no;
+	bool sync;
+	bool priority;
+	bool nop_last;
+	bool pre_delay;
+	bool post_delay;
+	int pkt_len;
+	int arb_lost_retry;
+	bool ch_last;
+	bool ccu_last;
+	int frame_cookie;
+};
+
+/*
+ * struct ccu_i2c_config - represents ccu configurations.
+ * @msg: pointer to struct ccu_msg_config
+ * @idx: indicates the index for gpii
+ * @mode: indicated CDM(1) or SYNC(0) mode
+ * @inter_delay: inter transfer delay
+ * @gpi: handle for struct ccu_gpii_info
+ */
+struct ccu_i2c_config {
+	struct ccu_msg_config *msg[MAX_NUM_MSGS];
+	int idx; /* Represents the GPII number */
+	int mode; /* CDM(1) or SYNC(0) mode */
+	unsigned long inter_delay;
+	struct ccu_gpii_info gpi[CCU_MAX_GPII];
+};
 
 struct geni_i2c_dev {
 	struct geni_se se;
@@ -119,13 +196,15 @@ struct geni_i2c_dev {
 	void *dma_buf;
 	size_t xfer_len;
 	dma_addr_t dma_addr;
-	struct dma_chan *tx_c;
-	struct dma_chan *rx_c;
+	struct dma_chan *tx_c[CCU_MAX_GPII];
+	struct dma_chan *rx_c[CCU_MAX_GPII];
 	bool no_dma;
 	bool abort_done;
 	bool is_shared;
 	const struct geni_i2c_desc *dev_data;
 	enum geni_se_xfer_mode xfer_mode;
+	bool is_ccu_qup;
+	struct ccu_i2c_config ccu;
 };
 
 struct geni_i2c_err_log {
@@ -133,6 +212,8 @@ struct geni_i2c_err_log {
 	const char *msg;
 };
 
+static int geni_i2c_runtime_resume(struct device *dev);
+
 static const struct geni_i2c_err_log gi2c_log[] = {
 	[GP_IRQ0] = {-EIO, "Unknown I2C err GP_IRQ0"},
 	[NACK] = {-ENXIO, "NACK: slv unresponsive, check its power/reset-ln"},
@@ -559,13 +640,16 @@ static void geni_i2c_gpi_unmap(struct geni_i2c_dev *gi2c, struct i2c_msg *msg,
 			       void *tx_buf, dma_addr_t tx_addr,
 			       void *rx_buf, dma_addr_t rx_addr)
 {
+	u16 len;
+
+	len = gi2c->is_ccu_qup ? (msg->len - CCU_DATA_OFFSET) : msg->len;
 	if (tx_buf) {
-		dma_unmap_single(gi2c->se.dev->parent, tx_addr, msg->len, DMA_TO_DEVICE);
+		dma_unmap_single(gi2c->se.dev->parent, tx_addr, len, DMA_TO_DEVICE);
 		i2c_put_dma_safe_msg_buf(tx_buf, msg, !gi2c->err);
 	}
 
 	if (rx_buf) {
-		dma_unmap_single(gi2c->se.dev->parent, rx_addr, msg->len, DMA_FROM_DEVICE);
+		dma_unmap_single(gi2c->se.dev->parent, rx_addr, len, DMA_FROM_DEVICE);
 		i2c_put_dma_safe_msg_buf(rx_buf, msg, !gi2c->err);
 	}
 }
@@ -574,7 +658,7 @@ static int geni_i2c_gpi(struct geni_i2c_dev *gi2c, struct i2c_msg *msg,
 			struct dma_slave_config *config, dma_addr_t *dma_addr_p,
 			void **buf, unsigned int op, struct dma_chan *dma_chan)
 {
-	struct gpi_i2c_config *peripheral;
+	struct gpi_i2c_config *peripheral_cfg;
 	unsigned int flags;
 	void *dma_buf;
 	dma_addr_t addr;
@@ -583,26 +667,39 @@ static int geni_i2c_gpi(struct geni_i2c_dev *gi2c, struct i2c_msg *msg,
 	struct dma_async_tx_descriptor *desc;
 	int ret;
 
-	peripheral = config->peripheral_config;
+	u16 len = gi2c->is_ccu_qup ? (msg->len - CCU_DATA_OFFSET) : msg->len;
 
-	dma_buf = i2c_get_dma_safe_msg_buf(msg, 1);
-	if (!dma_buf)
-		return -ENOMEM;
+	peripheral_cfg = config->peripheral_config;
+	if (gi2c->is_ccu_qup) {
+		if (op == I2C_WRITE)
+			dma_buf = kmemdup((msg->buf + CCU_DATA_OFFSET), len, GFP_KERNEL);
+		else
+			dma_buf = (msg->buf + CCU_DATA_OFFSET);
+
+		if (!dma_buf)
+			return -ENOMEM;
+		i2c_trace_log(gi2c->se.dev, "%s: CCU QUP - msg.len:%d - [%d]\n",
+			      __func__, msg->len, len);
+	} else {
+		dma_buf = i2c_get_dma_safe_msg_buf(msg, 1);
+		if (!dma_buf)
+			return -ENOMEM;
+		i2c_trace_log(gi2c->se.dev, "%s: QUP - msg.len:%d\n", __func__, msg->len);
+	}
 
 	if (op == I2C_WRITE)
 		map_dirn = DMA_TO_DEVICE;
 	else
 		map_dirn = DMA_FROM_DEVICE;
 
-	addr = dma_map_single(gi2c->se.dev->parent, dma_buf, msg->len, map_dirn);
+	addr = dma_map_single(gi2c->se.dev->parent, dma_buf, len, map_dirn);
 	if (dma_mapping_error(gi2c->se.dev->parent, addr)) {
 		i2c_put_dma_safe_msg_buf(dma_buf, msg, false);
 		return -ENOMEM;
 	}
 
-	/* set the length as message for rx txn */
-	peripheral->rx_len = msg->len;
-	peripheral->op = op;
+	peripheral_cfg->rx_len = len;
+	peripheral_cfg->op = op;
 
 	ret = dmaengine_slave_config(dma_chan, config);
 	if (ret) {
@@ -611,8 +708,9 @@ static int geni_i2c_gpi(struct geni_i2c_dev *gi2c, struct i2c_msg *msg,
 		goto err_config;
 	}
 
-	peripheral->set_config = 0;
-	peripheral->multi_msg = true;
+	peripheral_cfg->set_config = 0;
+	peripheral_cfg->multi_msg = true;
+
 	flags = DMA_PREP_INTERRUPT | DMA_CTRL_ACK;
 
 	if (op == I2C_WRITE)
@@ -620,7 +718,7 @@ static int geni_i2c_gpi(struct geni_i2c_dev *gi2c, struct i2c_msg *msg,
 	else
 		dma_dirn = DMA_DEV_TO_MEM;
 
-	desc = dmaengine_prep_slave_single(dma_chan, addr, msg->len, dma_dirn, flags);
+	desc = dmaengine_prep_slave_single(dma_chan, addr, len, dma_dirn, flags);
 	if (!desc) {
 		I2C_ERR_LOG(LOG_LVL_ERROR, gi2c->se.dev,
 			    "prep_slave_sg failed\n");
@@ -629,7 +727,7 @@ static int geni_i2c_gpi(struct geni_i2c_dev *gi2c, struct i2c_msg *msg,
 	}
 
 	desc->callback_result = i2c_gpi_cb_result;
-	desc->callback_param = peripheral;
+	desc->callback_param = peripheral_cfg;
 
 	dmaengine_submit(desc);
 	*buf = dma_buf;
@@ -638,16 +736,88 @@ static int geni_i2c_gpi(struct geni_i2c_dev *gi2c, struct i2c_msg *msg,
 	return 0;
 
 err_config:
-	dma_unmap_single(gi2c->se.dev->parent, addr, msg->len, map_dirn);
-	i2c_put_dma_safe_msg_buf(dma_buf, msg, false);
+	dma_unmap_single(gi2c->se.dev->parent, addr, len, map_dirn);
+	if (gi2c->is_ccu_qup)
+		kfree(dma_buf);
+	else
+		i2c_put_dma_safe_msg_buf(dma_buf, msg, false);
+
 	return ret;
 }
 
+/*
+ * geni_i2c_extract_frame_cfg - used to get the configuration of data xfer which is
+ * 				passed by client in msg buffer.
+ * @gi2c: points to the geni i2c dev structure
+ * @msg: points to the i2c msg structure
+ * @idx: index of message
+ *
+ * Returns 0 on success or error on failure
+ */
+static int geni_i2c_extract_frame_cfg(struct geni_i2c_dev *gi2c, struct i2c_msg *msg, int msg_idx)
+{
+	u64 data;
+	u64 *msg_buf = NULL;
+
+	if (msg->len <= CCU_DATA_OFFSET) {
+		dev_err(gi2c->se.dev, "%s: CCU config data not present. len:%d\n",
+			__func__, msg->len);
+		return -EINVAL;
+	}
+
+	msg_buf = kmemdup(msg->buf, CCU_DATA_OFFSET, GFP_KERNEL);
+	if (!msg_buf) {
+		dev_err(gi2c->se.dev, "%s:  Memory allocation failed\n", __func__);
+		return -ENOMEM;
+	}
+
+	gi2c->ccu.msg[msg_idx] = kzalloc(sizeof(struct ccu_msg_config), GFP_KERNEL);
+	if (!gi2c->ccu.msg[msg_idx]) {
+		dev_err(gi2c->se.dev, "%s:  Memory allocation failed for msg[%d]\n",
+			__func__, msg_idx);
+		kfree(msg_buf);
+		return -ENOMEM;
+	}
+	data = *msg_buf;
+
+	gi2c->ccu.msg[msg_idx]->gpii_no = CCU_GET_GPII(data);
+	gi2c->ccu.msg[msg_idx]->sync =  CCU_GET_SYNC(data);
+	gi2c->ccu.msg[msg_idx]->priority = CCU_GET_PRIO(data);
+	gi2c->ccu.msg[msg_idx]->nop_last = CCU_GET_NOP_FRAME(data);
+	gi2c->ccu.msg[msg_idx]->pre_delay = CCU_GET_INSERT_DELAY_BEFORE(data);
+	gi2c->ccu.msg[msg_idx]->post_delay = CCU_GET_INSERT_DELAY_AFTER(data);
+	gi2c->ccu.msg[msg_idx]->ch_last = CCU_GET_CH_LAST(data);
+	gi2c->ccu.msg[msg_idx]->ccu_last = CCU_GET_LAST(data);
+	gi2c->ccu.msg[msg_idx]->arb_lost_retry = CCU_GET_ARB_LOST_CNT(data);
+	gi2c->ccu.msg[msg_idx]->pkt_len = CCU_GET_PKT_LEN(data);
+	gi2c->ccu.msg[msg_idx]->frame_cookie = CCU_GET_FRAME_COOKIE(data);
+
+	i2c_trace_log(gi2c->se.dev, "%s: *** Message[%d] configuration ***\n"
+				  "GPII:%d Sync:%d Prio:%d Delay after:%d\n"
+				  "Delay Before:%d ch_last:%d ccu_last:%d\n"
+				  "arb_lost_retry:%d pkt_len:%d\n"
+				  "cookie:0x%x\n", __func__, msg_idx,
+				  gi2c->ccu.msg[msg_idx]->gpii_no,
+				  gi2c->ccu.msg[msg_idx]->sync,
+				  gi2c->ccu.msg[msg_idx]->priority,
+				  gi2c->ccu.msg[msg_idx]->post_delay,
+				  gi2c->ccu.msg[msg_idx]->pre_delay,
+				  gi2c->ccu.msg[msg_idx]->ch_last,
+				  gi2c->ccu.msg[msg_idx]->ccu_last,
+				  gi2c->ccu.msg[msg_idx]->arb_lost_retry,
+				  gi2c->ccu.msg[msg_idx]->pkt_len,
+				  gi2c->ccu.msg[msg_idx]->frame_cookie);
+	kfree(msg_buf);
+	msg_buf = NULL;
+
+	return 0;
+}
+
 static int geni_i2c_gpi_xfer(struct geni_i2c_dev *gi2c, struct i2c_msg msgs[], int num)
 {
 	struct dma_slave_config config = {};
 	struct gpi_i2c_config *peripheral;
-	int i, ret = 0;
+	int i, k, ret = 0, idx = 0;
 	unsigned long time_left;
 	dma_addr_t tx_addr, rx_addr;
 	void *tx_buf = NULL, *rx_buf = NULL;
@@ -682,6 +852,15 @@ static int geni_i2c_gpi_xfer(struct geni_i2c_dev *gi2c, struct i2c_msg msgs[], i
 
 		peripheral->first_msg = (i == 0) ? 1 : 0;
 		peripheral->last_msg = (i == num - 1) ? 1 : 0;
+
+		if (gi2c->is_ccu_qup) {
+			ret = geni_i2c_extract_frame_cfg(gi2c, gi2c->cur, i);
+			if (ret)
+				return ret;
+			idx = (gi2c->ccu.msg[i]->gpii_no % CCU_MAX_GPII);
+		}
+		i2c_trace_log(gi2c->se.dev, "msg[%d].len:%d GPII:%d\n", i, gi2c->cur->len, idx);
+
 		peripheral->stretch = 0;
 		if (i < num - 1)
 			peripheral->stretch = 1;
@@ -689,20 +868,20 @@ static int geni_i2c_gpi_xfer(struct geni_i2c_dev *gi2c, struct i2c_msg msgs[], i
 		peripheral->addr = msgs[i].addr;
 
 		ret =  geni_i2c_gpi(gi2c, &msgs[i], &config,
-				    &tx_addr, &tx_buf, I2C_WRITE, gi2c->tx_c);
+				    &tx_addr, &tx_buf, I2C_WRITE, gi2c->tx_c[idx]);
 		if (ret)
 			goto err;
 
 		if (msgs[i].flags & I2C_M_RD) {
 			ret =  geni_i2c_gpi(gi2c, &msgs[i], &config,
-					    &rx_addr, &rx_buf, I2C_READ, gi2c->rx_c);
+					    &rx_addr, &rx_buf, I2C_READ, gi2c->rx_c[idx]);
 			if (ret)
 				goto err;
 
-			dma_async_issue_pending(gi2c->rx_c);
+			dma_async_issue_pending(gi2c->rx_c[idx]);
 		}
 
-		dma_async_issue_pending(gi2c->tx_c);
+		dma_async_issue_pending(gi2c->tx_c[idx]);
 
 		time_left = wait_for_completion_timeout(&gi2c->done, XFER_TIMEOUT);
 		if (!time_left)
@@ -714,15 +893,24 @@ static int geni_i2c_gpi_xfer(struct geni_i2c_dev *gi2c, struct i2c_msg msgs[], i
 		}
 
 		geni_i2c_gpi_unmap(gi2c, &msgs[i], tx_buf, tx_addr, rx_buf, rx_addr);
+		if (gi2c->is_ccu_qup)
+			kfree(gi2c->ccu.msg[i]);
 	}
-
 	return num;
 
 err:
-	I2C_ERR_LOG(LOG_LVL_ERROR, gi2c->se.dev, "GPI transfer failed: %d\n",
-		    ret);
-	dmaengine_terminate_sync(gi2c->rx_c);
-	dmaengine_terminate_sync(gi2c->tx_c);
+	dev_err(gi2c->se.dev, "GPI transfer failed: %d\n", ret);
+	if (gi2c->is_ccu_qup) {
+		for (k = i; k >= 0; k--)
+			kfree(gi2c->ccu.msg[k]);
+		for (k = 0; k < CCU_MAX_GPII; k++) {
+			dmaengine_terminate_sync(gi2c->rx_c[k]);
+			dmaengine_terminate_sync(gi2c->tx_c[k]);
+		}
+	} else {
+		dmaengine_terminate_sync(gi2c->rx_c[idx]);
+		dmaengine_terminate_sync(gi2c->tx_c[idx]);
+	}
 	geni_i2c_gpi_unmap(gi2c, &msgs[i], tx_buf, tx_addr, rx_buf, rx_addr);
 	return ret;
 }
@@ -819,38 +1007,66 @@ MODULE_DEVICE_TABLE(acpi, geni_i2c_acpi_match);
 
 static void release_gpi_dma(struct geni_i2c_dev *gi2c)
 {
-	if (gi2c->rx_c)
-		dma_release_channel(gi2c->rx_c);
+	int i, idx;
 
-	if (gi2c->tx_c)
-		dma_release_channel(gi2c->tx_c);
+	idx =  gi2c->is_ccu_qup ? CCU_MAX_GPII : 1;
+
+	for (i = 0; i < idx; i++) {
+		if (gi2c->rx_c[i]) {
+			dma_release_channel(gi2c->rx_c[i]);
+			gi2c->rx_c[i] = NULL;
+		}
+		if (gi2c->tx_c[i]) {
+			dma_release_channel(gi2c->tx_c[i]);
+			gi2c->tx_c[i] = NULL;
+		}
+	}
 }
 
 static int setup_gpi_dma(struct geni_i2c_dev *gi2c)
 {
 	int ret;
+	int i = 0, idx = 0;
+	/* For TOP QUP, only a single GPII is used per serial engine. However, for CCU QUP,
+	 * up to three GPIIs can be used per serial engine. When a single GPII is used, the
+	 * "tx" and "rx" names are provided using the `dma-names` property from the DTSI.
+	 * To maintain backward compatibility with older targets, the same names are used,
+	 * with two additional names added for the second and third GPIIs.
+	 */
+	const char *tx_ch[CCU_MAX_GPII] = {"tx", "tx1", "tx2"};
+	const char *rx_ch[CCU_MAX_GPII] = {"rx", "rx1", "rx2"};
+
+	if (gi2c->is_ccu_qup)
+		idx = CCU_MAX_GPII;
 
 	geni_se_select_mode(&gi2c->se, GENI_GPI_DMA);
-	gi2c->tx_c = dma_request_chan(gi2c->se.dev, "tx");
-	if (IS_ERR(gi2c->tx_c)) {
-		ret = dev_err_probe(gi2c->se.dev, PTR_ERR(gi2c->tx_c),
-				    "Failed to get tx DMA ch\n");
-		goto err_tx;
-	}
+	do {
+		i2c_trace_log(gi2c->se.dev, "%s:[%d] Request TX chan:%s\n", __func__, i, tx_ch[i]);
+		gi2c->tx_c[i] = dma_request_chan(gi2c->se.dev, tx_ch[i]);
+		if (IS_ERR(gi2c->tx_c[i])) {
+			ret = dev_err_probe(gi2c->se.dev, PTR_ERR(gi2c->tx_c[i]),
+					    "Failed to get tx%d DMA ch\n", i);
+			goto err;
+		}
 
-	gi2c->rx_c = dma_request_chan(gi2c->se.dev, "rx");
-	if (IS_ERR(gi2c->rx_c)) {
-		ret = dev_err_probe(gi2c->se.dev, PTR_ERR(gi2c->rx_c),
-				    "Failed to get rx DMA ch\n");
-		goto err_rx;
-	}
+		i2c_trace_log(gi2c->se.dev, "%s:[%d] Request RX chan:%s\n", __func__, i, rx_ch[i]);
+		gi2c->rx_c[i] = dma_request_chan(gi2c->se.dev, rx_ch[i]);
+		if (IS_ERR(gi2c->rx_c[i])) {
+			ret = dev_err_probe(gi2c->se.dev, PTR_ERR(gi2c->rx_c[i]),
+					    "Failed to get rx%d DMA ch\n", i);
+			dma_release_channel(gi2c->tx_c[i]);
+			goto err;
+		}
+	} while (++i < idx);
 
-	I2C_DBG_LOG(LOG_LVL_INFO, gi2c->se.dev, "Grabbed GPI dma channels\n");
+	i2c_trace_log(gi2c->se.dev, "Grabbed GPI dma channels\n");
 	return 0;
 
-err_rx:
-	dma_release_channel(gi2c->tx_c);
-err_tx:
+err:
+	for (idx = i; idx > 0; idx--) {
+		dma_release_channel(gi2c->tx_c[idx - 1]);
+		dma_release_channel(gi2c->rx_c[idx - 1]);
+	}
 	return ret;
 }
 
@@ -1008,7 +1224,8 @@ static int geni_i2c_probe(struct platform_device *pdev)
 
 	if (!(of_device_is_compatible(dev->of_node, "qcom,sa8255p-geni-i2c") ||
 			of_device_is_compatible(dev->of_node, "qcom,sa7255p-geni-i2c") ||
-			of_device_is_compatible(dev->of_node, "qcom,sa8797p-geni-i2c"))) {
+			of_device_is_compatible(dev->of_node, "qcom,sa8797p-geni-i2c") ||
+			of_device_is_compatible(dev->of_node, "qcom,sa8797p-geni-i2c-ccu"))) {
 		if (desc->has_core_clk) {
 			gi2c->core_clk = devm_clk_get(dev, "core");
 			if (IS_ERR(gi2c->core_clk))
@@ -1045,6 +1262,11 @@ static int geni_i2c_probe(struct platform_device *pdev)
 		gi2c->clk_freq_out = KHZ(100);
 	}
 
+	if (desc && desc->is_ccu_qup) {
+		gi2c->is_ccu_qup = true;
+		i2c_trace_log(&pdev->dev, "%s:CCU QUP enabled\n", __func__);
+	}
+
 	if (has_acpi_companion(dev))
 		ACPI_COMPANION_SET(&gi2c->adap.dev, ACPI_COMPANION(dev));
 
@@ -1211,11 +1433,19 @@ static const struct dev_pm_ops geni_i2c_pm_ops = {
 static const struct geni_i2c_desc geni_i2c = {
 	.icc_ddr = "qup-memory",
 	.geni_se_switch_state = geni_i2c_switch_resource_state,
+	.is_ccu_qup = false,
+};
+
+static const struct geni_i2c_desc remotely_geni_i2c_ccu = {
+	.icc_ddr = "qup-memory",
+	.geni_se_switch_state = NULL,
+	.is_ccu_qup = true,
 };
 
 static const struct geni_i2c_desc remotely_geni_i2c = {
 	.icc_ddr = "qup-memory",
 	.geni_se_switch_state = NULL,
+	.is_ccu_qup = false,
 };
 
 static const struct geni_i2c_desc i2c_master_hub = {
@@ -1224,6 +1454,7 @@ static const struct geni_i2c_desc i2c_master_hub = {
 	.no_dma_support = true,
 	.tx_fifo_depth = 16,
 	.geni_se_switch_state = geni_i2c_switch_resource_state,
+	.is_ccu_qup = false,
 };
 
 static const struct of_device_id geni_i2c_dt_match[] = {
@@ -1232,6 +1463,7 @@ static const struct of_device_id geni_i2c_dt_match[] = {
 	{ .compatible = "qcom,sa7255p-geni-i2c", .data = &remotely_geni_i2c },
 	{ .compatible = "qcom,sa8255p-geni-i2c", .data = &remotely_geni_i2c },
 	{ .compatible = "qcom,sa8797p-geni-i2c", .data = &remotely_geni_i2c },
+	{ .compatible = "qcom,sa8797p-geni-i2c-ccu", .data = &remotely_geni_i2c_ccu },
 	{}
 };
 MODULE_DEVICE_TABLE(of, geni_i2c_dt_match);
-- 
2.34.1

